src/constructors/lattices.c

```diff
@@ -0,0 +1,308 @@
+            igraph_vector_int_push_back(&edges, VERTEX_INDEX((i), (j)));                                                                          \
+        }                                                                                                                                         \
+    }
+
+     /* computing the number of edges in the constructed triangle lattice */
+    igraph_integer_t no_of_edges_2 = VECTOR(*row_lengths_vector)[row_count - 1] - 1;
+    igraph_integer_t multiplier = mutual && directed ? 4 : 2;
+    for (j = 0; j < row_count - 1; j++)
+    {
+        IGRAPH_SAFE_ADD(no_of_edges_2, VECTOR(*row_lengths_vector)[j], &no_of_edges_2);
+        IGRAPH_SAFE_ADD(no_of_edges_2, fmin(ROW_END(j), ROW_END((j + 1))) - fmax(VECTOR(*row_start_vector)[j], VECTOR(*row_start_vector)[j + 1]) + 1,
+                        &no_of_edges_2);
+        IGRAPH_SAFE_ADD(no_of_edges_2, fmin(ROW_END(j), ROW_END((j + 1)) + 1) - fmax(VECTOR(*row_start_vector)[j], VECTOR(*row_start_vector)[j + 1] + 1) + 1,
+                        &no_of_edges_2);
+    }
+    IGRAPH_SAFE_MULT(no_of_edges_2, multiplier, &no_of_edges_2);
+    IGRAPH_CHECK(igraph_vector_int_reserve(&edges, no_of_edges_2));
+
+    /* constructing the edge array */
+    igraph_integer_t k;
+    for (j = 0; j < row_count; j++)
+    {
+        IGRAPH_ALLOW_INTERRUPTION();
+        for (i = 0; i < VECTOR(*row_lengths_vector)[j]; i++)
+        {
+            k = VECTOR(*row_start_vector)[j] + i;
+            ADD_EDGE_IJ_KL_IF_EXISTS(k, j, (k + 1), j);
+            if (j < row_count - 1)
+            {
+                ADD_EDGE_IJ_KL_IF_EXISTS(k, j, k, (j + 1));
+                ADD_EDGE_IJ_KL_IF_EXISTS(k, j, (k - 1), (j + 1));
+            }
+        }
+    }
+
+    IGRAPH_CHECK(igraph_create(graph, &edges, no_of_nodes, directed));
+
+    igraph_vector_int_destroy(&row_lengths_prefix_sum_vector);
+    igraph_vector_int_destroy(&edges);
+    IGRAPH_FINALLY_CLEAN(2);
+
+    return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t triangle_lattice_triangle_shape(igraph_t *graph, igraph_integer_t size, igraph_bool_t directed, igraph_bool_t mutual)
+{
+    igraph_integer_t row_count = size;
+    igraph_vector_int_t row_lengths_vector;
+    igraph_vector_int_t row_start_vector;
+    igraph_integer_t i;
+
+    IGRAPH_CHECK(igraph_vector_int_init(&row_lengths_vector, row_count));
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &row_lengths_vector);
+    IGRAPH_CHECK(igraph_vector_int_init(&row_start_vector, row_count));
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &row_start_vector);
+
+    for (i = 0; i < row_count; i++)
+    {
+        VECTOR(row_lengths_vector)[i] = size - i;
+        VECTOR(row_start_vector)[i] = 0;
+    }
+
+    IGRAPH_CHECK(triangle_lattice(graph, directed, mutual, &row_lengths_vector, &row_start_vector));
+
+    igraph_vector_int_destroy(&row_lengths_vector);
+    IGRAPH_FINALLY_CLEAN(1);
+    igraph_vector_int_destroy(&row_start_vector);
+    IGRAPH_FINALLY_CLEAN(1);
```

Comment:
The first `IGRAPH_FINALLY_CLEAN(...)` is not necessary and the second one should be `IGRAPH_FINALLY_CLEAN(2)`.